src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJF.java

```diff
@@ -0,0 +1,615 @@
+        private long read = 0;
+
+        private FileBlock block;
+
+        private boolean closed = false;
+
+        private final int jointIdx;
+
+        private final String originalThreadName;
+
+        private final long ctorTime;
+
+        /**
+         * Map the content already written to disk for reading. If the flushed count exceeds MAX_BUFFER_SIZE, use the max instead.
+         */
+        JoinInputStream( int jointIdx ) throws IOException
+        {
+            this.jointIdx = jointIdx;
+            block = metadata.getFirstBlock();
+            this.originalThreadName = Thread.currentThread().getName();
+            this.ctorTime = System.nanoTime();
+        }
+
+        @Override
+        public boolean equals( final Object o )
+        {
+            if ( this == o )
+            {
+                return true;
+            }
+            if ( !( o instanceof JoinInputStream ) )
+            {
+                return false;
+            }
+
+            final JoinInputStream that = (JoinInputStream) o;
+
+            if ( jointIdx != that.jointIdx )
+            {
+                return false;
+            }
+            if ( ctorTime != that.ctorTime )
+            {
+                return false;
+            }
+            return originalThreadName.equals( that.originalThreadName );
+        }
+
+        @Override
+        public int hashCode()
+        {
+            int result = jointIdx;
+            result = 31 * result + originalThreadName.hashCode();
+            result = 31 * result + (int) ( ctorTime ^ ( ctorTime >>> 32 ) );
+            return result;
+        }
+
+        /**
+         * If this stream is in the process of closing, throw {@link IOException}. While the read-bytes count in this
+         * stream equals the flushed-bytes count in the associated output stream, wait for new content. If the output stream closes while we're
+         * waiting, return -1. If the thread is interrupted while we're waiting, return -1.
+         *
+         * If the current mapped buffer has been completely read, map the next section of content from the file. Then read the next byte from the
+         * buffer, increment the read-bytes count, and return it.
+         */
+        @Override
+        public int read() throws IOException
+        {
+            synchronized ( InfinispanJF.this )
+            {
+                if ( closed )
+                {
+                    throw new IOException( "Joint: " + jointIdx + "(" + originalThreadName
+                                                           + "): Cannot read from closed stream!" );
+                }
+
+            }
+            // Is it better to write getting logic into FileBlock rather than expose the whole buffer?
+            if ( block.getBuffer().position() == block.getBuffer().limit() )
+            {
+                try
+                {
+                    FileBlock next = filesystem.getNextBlock( block );
+                    block = next;
```

Comment:
We need a null check here. If next is null, that means we hit EOF, I think?

The other alternative would be to return the same block again, maybe, and let the EOF check below trip.

Reply:
I'm going to move the EOF check up so that we only check it when the buffer has already been read - I hadn't thought of that before but the way it is there if the new block is EOF but has some data in it read would return -1 even though there's still data. I'll add a null check in addition to this.